python/shotgun_model/shotgun_hierarchy_model.py

```diff
@@ -0,0 +1,855 @@
+        logger.debug("Path: %s" % (self._path,))
+        logger.debug("Seed entity field: %s" % (self._seed_entity_field,))
+        logger.debug("Entity fields: %s" % (self._entity_fields,))
+
+        logger.debug("First population pass: Calling _load_external_data()")
+        self._load_external_data()
+        logger.debug("External data population done.")
+
+        # only one column. give it a default value
+        self.setHorizontalHeaderLabels(
+            ["%s Hierarchy" % (self._seed_entity_field,)]
+        )
+
+        return self._load_cached_data()
+
+    def _on_data_retriever_work_failure(self, uid, msg):
+        """
+        Asynchronous callback - the data retriever failed to do some work
+
+        :param uid: The unique id of the work that failed
+        :param msg: The error message returned for the failure
+        """
+        uid = sanitize_qt(uid)  # qstring on pyqt, str on pyside
+        msg = sanitize_qt(msg)
+
+        if uid not in self._running_query_lookup:
+            # not our job. ignore
+            logger.debug("Retrieved error from data worker: %s" % (msg,))
+            return
+
+        url = self._running_query_lookup[uid]
+
+        # query is done. clear it out
+        del self._running_query_lookup[uid]
+
+        full_msg = (
+            "Error retrieving data from Shotgun.\n"
+            "  URL: %s\n"
+            "  Error: %s\n"
+        ) % (url, msg)
+        self.data_refresh_fail.emit(full_msg)
+        logger.warning(full_msg)
+
+    def _on_data_retriever_work_completed(self, uid, request_type, data):
+        """
+        Signaled whenever the data retriever completes some work.
+        This method will dispatch the work to different methods
+        depending on what async task has completed.
+
+        :param uid:             The unique id of the work that completed
+        :param request_type:    Type of work completed
+        :param data:            Result of the work
+        """
+        uid = sanitize_qt(uid)  # qstring on pyqt, str on pyside
+        data = sanitize_qt(data)
+
+        logger.debug("Received worker payload of type: %s" % (request_type,))
+
+        if uid not in self._running_query_lookup:
+            # not our job. ignore.
+            return
+
+        # query is done. clear it out
+        del self._running_query_lookup[uid]
+
+        nav_data = data["nav"]
+        self.__on_nav_data_arrived(nav_data)
+
+    def _populate_default_thumbnail(self, item):
+        """
+        Sets the icon for the supplied item based on its "kind" as returned
+        by the ``nav_expand()`` api call.
+
+        :param item: The :class:`~PySide.QtGui.QStandardItem` item to set the
+            icon for.
+        """
+
+        data = get_sg_data(item)
+
+        item_ref = data.get("ref", {})
+        item_kind = item_ref.get("kind")
+
+        if item_kind == "entity":
+            entity_type = item_ref.get("value", {}).get("type")
```

Comment:
What if `item_ref["value"]` is `None`? does that never happen?
Maybe `(item_ref.get("value") or {}).get("type")` instead?